### ISSUE TITLE
data: image not blocked

### ISSUE THREAD

**Issue by chiraag-nataraj** [uBlockOrigin/uMatrix-issues#141]
### Prerequisites

- [x] I performed a cursory search of the issue tracker to avoid opening a duplicate issue
    - Your issue may already be reported.
    - [x] I also searched the existing issues at <https://github.com/gorhill/uMatrix/issues>
- [x] This is not a support issue or a question
    - Support issues and questions are handled at [/r/uMatrix](https://old.reddit.com/r/uMatrix/)
- I tried to reproduce the issue when...
    - [x] uMatrix is the only extension
    - [ ] uMatrix with default lists/settings
    - [x] using a new, unmodified browser profile
- [x] I am running the latest version of uMatrix
- [x] I checked the [documentation](https://github.com/gorhill/uMatrix/wiki) to understand that the issue I report is not a normal behavior
- [x] I used the logger to rule out that the issue is caused by my ruleset

### Description

It seems as though inline `data:` URLs are not blocked by disabling images from loading.

### A specific URL where the issue occurs

https://resources.whitesourcesoftware.com/blog-whitesource/top-5-linux-kernel-vulnerabilities-in-2018

### Steps to Reproduce

1. Ruleset: Only 1st-party CSS and CSS from `cdnwtrk.com` allowed. Everything else should be disabled.
2. Note that the image next to "Get 451 Research's Report" is displayed.
3. Use "Inspect Element" to figure out that they're injecting the image through CSS using the `::after` selector, `background-image`, and a `data:` URL.

### Ruleset

See above. My default ruleset is to _only_ allow 1st-party CSS. I added `cdnwtrk.com` temporarily for this website to make it more readable.

### Supporting evidence

![2019-04-30-090326_3840x2160_scrot](https://user-images.githubusercontent.com/14947779/56963422-6b170980-6b48-11e9-9b83-ac488e61088b.png)

### Your environment

* uMatrix version: 1.3.16
* Browser Name and version: Firefox 66.0.1
* Operating System and version: Debian sid/experimental

**Comment by uBlock-user** [comment]
data: URI's are inline embedded files in the document itself, hence they cannot be blocked by rules.

**Comment by gorhill** [comment]
Duplicate of <https://github.com/gorhill/uMatrix/issues/378>.

**Comment by chiraag-nataraj** [comment]
> Edit: blocking first party CSS might work in your favour

That does indeed solve the problem (I actually tested that right after submitting the issue).

@gorhill Given your feedback on the other bug, I presume this is something that will not be addressed? If so, I'll go ahead and close this.

**Comment by uBlock-user** [comment]
> I presume this is something that will not be addressed

There's nothing to address, uMatrix only blocks things that make network connections to fetch something, there's no such thing happening here, so nothing to block.

**Comment by gorhill** [comment]
The images could be prevented from being rendered by the browser with a CSP directive `img-src`, the question is whether it's really worth it -- so far I am not convinced it is.

**Comment by uBlock-user** [comment]
> The images could be prevented from being rendered by the browser with a CSP directive img-src,

That can be used when a user blocks images image rule, actually that should be inserted when the blocks all images via `* * image block`

**Comment by chiraag-nataraj** [comment]
@gorhill @uBlock-user That would be perfect, actually.

**Comment by gorhill** [comment]
What is the benefit of preventing the browser of rendering what has already been downloaded? It makes sense to prevent execution of inline JS, but does it for inline images?

**Comment by uBlock-user** [comment]
or when a user wants to block all images from a specific website -- `github.com * image block`

**Comment by uBlock-user** [comment]
> What is the benefit of preventing the browser from rendering what has already been downloaded

Not having to rely on external tools to block stuff rendered from data: URIs I guess, in this case blocking of  first party CSS works without breakage, but that will not work on other websites and uMatrix is pretty much helpless at the moment when it comes to blocking stuff from data:URIs.

**Comment by gorhill** [comment]
> Not having to rely on external tools to block stuff rendered from data: URIs

That does not answer the question: Why prevent rendering of in the first place? What is to be gained? Let's say the CSS workaround does work here, why is it so bad that the image is being rendered in the current case? (beside "I do not want to see it") There is no secondary network requests possible by rendering inlined images, so that shouldn't be seen as an issue.

**Comment by uBlock-user** [comment]
> why is it so bad that the image is being rendered in the current case? (beside "I do not want to see it") 

It isn't bad, but when I **block** all images, I expect all images to be **blocked** including the ones rendered by `data:`URIs, other than this, no other reason for me.

**Comment by gorhill** [comment]
Academic arguments are not worth the real world added complexity. There is no connection established to download `data:` URI-based images, they will never show up in the reported count in matrix or the logger. Someone blocking all images wouldn't know that a page has images if they were all `data:` URI-based. uMatrix is first a tool to control network requests, not a tool to implement cosmetic preferences. Since there is no sensible benefit for not rendering `data:` URI-based images (unlike JS/CSS), I prefer to not spend time on this (and that would be at the expense of other efforts).

**Comment by uBlock-user** [comment]
> There is no connection established to download data: URI-based images, they will never show up in the reported count in matrix or the logger. Someone blocking all images wouldn't know that a page has images if they were all data: URI-based. uMatrix is first a tool to control network requests, not a tool to implement cosmetic preferences.

True. Yes, I know all of that, but since you questioned and entertained the possibility of using CSP, I answered for the sake of answering. Besides, I do have uBO installed, so I always have a way to block that banner completely.
